### PR TITLE
[CAS] Add caching diagnostic processor

### PR DESCRIPTION
**Pull request by cachemeifyoucan**
Add a CachedDiagnosticsProcessor that is a DiagConsumer can capture all the diagnostics during a compilation, serialized them into CAS with a format that can be replayed without re-compiling.

This is also tested via default all the diagnostics to capture and replayed via CachedDiagnosticsProcessor and pass all the tests.

### CODE REVIEW

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

File: include/swift/Frontend/CachedDiagnostics.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+class SourceManager;
+class FrontendInputsAndOutputs;
+
+class CachingDiagnosticsProcessor {
+public:
+  CachingDiagnosticsProcessor(CompilerInstance &Instance);
+  ~CachingDiagnosticsProcessor();
+
+  /// Start capturing all the diagnostics from DiagnosticsEngine.
+  void startCaptureDiagnostics();
```
Comment (benlangmuir) [inline include/swift/Frontend/CachedDiagnostics.h:36]: Naming: I suggest
* startDiagnosticCapture
* endDiagnosticCapture

File: include/swift/Frontend/CachedDiagnostics.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  /// End capturing all the diagnostics from DiagnosticsEngine.
+  void endCaptureDiagnostics();
+
+  /// Disable catpure.
+  void setDisableCapture();
```
Comment (benlangmuir) [inline include/swift/Frontend/CachedDiagnostics.h:41]: I don't understand what this does compared to the above start/end API.  From the implementation I can only disable capture if it's not already started, so what exactly does it mean to disable if it's not capturing?

**Review by benlangmuir — COMMENTED** [review]
> This is also tested via default all the diagnostics to capture and replayed via CachedDiagnosticsProcessor and pass all the tests.

To clarify, is this something that is always happening, or something you did manually?

**Comment by cachemeifyoucan** [comment]
> > This is also tested via default all the diagnostics to capture and replayed via CachedDiagnosticsProcessor and pass all the tests.
> 
> To clarify, is this something that is always happening, or something you did manually?

Manually, we can also leave a switch in there for testing purpose. There isn't a good way to dup all tests with a different configuration to have this always on and probably very expensive.

File: include/swift/Frontend/CachedDiagnostics.h (hunk 2 above)
Comment (cachemeifyoucan) [reply include/swift/Frontend/CachedDiagnostics.h:41]: Disable means do not store the output and throw away the capture state. This is for replaying diagnostics.

An empty capture (or never started capture) is no diagnostics and will be stored as no diagnostics in the CAS. While a disable capture doesn't trigger store.

Comment (benlangmuir) [reply include/swift/Frontend/CachedDiagnostics.h:41]: Would it make sense for replay to be a `static` method instead? Then you don't need to worry about the capturing "state" in the public API

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
ping

**Review by benlangmuir — APPROVED** [review]

**Review by artemcm — APPROVED** [review]

**Pull request merged**